### PR TITLE
PSG-679: Add psga-backend pipeline metadata file

### PR DESCRIPTION
**Pull request by Aiky30**
https://jira.congenica.net/browse/PSG-679

The configuration file should contain information about the fields the pipeline expects to be passed to its metadata file.
i.e. 
    - Sample ID
    - File Path
    - Kit

This configuration file should be part of the pipeline because:
    - We might end up with 100 pipelines, and all developed by different people.
    - The metadata will be pipeline version specific, so it should be updated by the pipeline developer when a new version is released to accurately reflect the pipeline. Therefore it doesn’t make sense to live in the BE, which is a separate project and should be entirely decoupled from the pipeline and it’s config.

### CODE REVIEW

File: psga/config/pipeline-schema.yaml [hunk 1]
```diff
@@ -0,0 +1,95 @@
+pipeline:
+  name: sars-cov-2
+  docker_image: ''
+  protocols:
+  - name: illumina-sars-cov-2-artic-v3
+    kit:
+      name: Artic V3
```
Comment (Aiky30) [inline psga/config/pipeline-schema.yaml:7]: @pdp10 You mentioned that these values are wrong. Can you please suggest the correct value. I believe it was something along the lines of: 

```suggestion
      name: Artic_V3
```

File: psga/config/pipeline-schema.yaml [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+      name: ONT
+      slug: ont
+    pathogen:
+      name: Sars-CoV-2
+      slug: sars-cov-2
+
+pipeline_version:
+  docker_tag: 1.0.1
+
+input_fields:
```
Comment (Aiky30) [inline psga/config/pipeline-schema.yaml:29]: @pdp10 We also need help defining the input and output fields that the pipeline expects to receive and output.

**Review by pdp10 — COMMENTED** [review]
Okay, I understand the reasons why this config should be in the pipeline.

@cramshaw I have a question though
Should not we have a pipeline-scheme.yaml per pathogen?
This would avoid handling with conflicts every time pipelines for different pathogens are edited. 
So I wonder whether a structure like: 
```
config 
    - sars_cov_2
          - pipeline-schema.yaml
    - tb
          - pipeline-schema.yaml
...
```
could simplify things for developers / reviewers.

File: psga/config/pipeline-schema.yaml (hunk 1 above)
Comment (Aiky30) [reply psga/config/pipeline-schema.yaml:7]: @pdp10 that's correct, all of the values that we are storing are for the purpose of being able to call the pipeline.

Comment (pdp10) [reply psga/config/pipeline-schema.yaml:7]: Values can be (case sensitive):

* `ARTIC_V${n}` with n in `["1", "2", "3", "4", "4.1"]`
* `Midnight-IDT_V1`
* `Midnight-ONT_V${n}` with n in `["1", "2", "3"]`
* `NEB-VarSkip_v${n}` with n in `["1a", "1a-long", "2", "2b"]`    (note that `v` is lower case here)

File: psga/config/pipeline-schema.yaml [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+      name: Illumina
+      slug: illumina
+    pathogen:
+      name: Sars-CoV-2
+      slug: sars-cov-2
+  - name: ont-sars-cov-2-artic-v4
+    kit:
+      name: Artic V4
+      slug: articv4
+    sequencer_technology:
```
Comment (pdp10) [inline psga/config/pipeline-schema.yaml:19]: This parameter is called `sequencing_technology` in the pipeline, in agreement to the architecture documents.
Values are (case sensitive: `illumina`, `ont`, `unknown`).

File: psga/config/pipeline-schema.yaml (hunk 2 above)
Comment (pdp10) [reply psga/config/pipeline-schema.yaml:29]: Input
* `sample_id` is a GUID. 
* I believe what here you call `batch` should be `run`. That's the parameter the pipeline uses. It says that that's a number. I would make that a GUID too.
* `study_type`, `score` and `collection_date` are not used by the pipeline

File: psga/config/pipeline-schema.yaml [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  validation_rules:
+  - rule_type: in_past
+    allow_today: false
+  - rule_type: not_null
+
+input_row_validation_rules:
+- rule_type: not_null
+  fields:
+  - collection_date
+  - lab_received_date
```
Comment (pdp10) [inline psga/config/pipeline-schema.yaml:72]: The pipeline does not use these.

File: psga/config/pipeline-schema.yaml [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  min: 1
+input_set_validation_rules:
+- rule_type: match_row_count
+  fields:
+  - control_sample
```
Comment (pdp10) [inline psga/config/pipeline-schema.yaml:77]: Not used.

**Review by pdp10 — COMMENTED** [review]
Added a few comments.

**Comment by Aiky30** [comment]
> Added a few comments.

Awesome thank you @pdp10. I'll assess the feedback and address where needed. I guess there is some cross over of the values in the file here where we need validation rules for a pipeline that actually have nothing to do with the pipeline. As this is the psga-pipeline package I guess this is where the 3rd party pipelines meet the psga requirements.

**Comment by Aiky30** [comment]
> Okay, I understand the reasons why this config should be in the pipeline.
> 
> @cramshaw I have a question though Should not we have a pipeline-scheme.yaml per pathogen? This would avoid handling with conflicts every time pipelines for different pathogens are edited. So I wonder whether a structure like:
> 
> ```
> config 
>     - sars_cov_2_pipeline_schema.yaml
>     - tb_pipeline_schema.yaml
> ...
> ```
> 
> could simplify things for developers / reviewers.

@pdp10, I think this would need to be pathogen and pipeline version. 

Maybe it is:
 `config/pathogen-name/pipeline-name-and-pipeline-version` i.e. `config/sars-cov-2/sars-cov-2-1.0.1.yaml`

Or even simpler 
Maybe it is `config/pathogen-name/pipeline-version` i.e. `config/sars-cov-2/1.0.1.yaml`

File: psga/config/pipeline-schema.yaml [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+  - '-'
+  min: 3
+  max: 5
+
+output_fields:
+- field_name: sample_id
+  field_type: string
+  required: true
+  validation_rules:
+  - {}
```
Comment (cramshaw) [inline psga/config/pipeline-schema.yaml:89]: We need to populate all the pipeline output fields here - @pdp10 can you link to the latest doc of the potential outputs? Are they in a confluence page? @Aiky30 we only need loose valdiation - data type and enums where necesssary

File: psga/config/pipeline-schema.yaml (hunk 5 above)
Comment (cramshaw) [reply psga/config/pipeline-schema.yaml:77]: It _may_ become required but we should drop for now.

File: psga/config/pipeline-schema.yaml (hunk 4 above)
Comment (cramshaw) [reply psga/config/pipeline-schema.yaml:72]: For now there won't be row or set validation rules on the pipeline.

File: psga/config/pipeline-schema.yaml (hunk 2 above)
Comment (cramshaw) [reply psga/config/pipeline-schema.yaml:29]: some of these field names will need to map to the BE and potentially the lab csv file.
I believe https://confluence.congenica.net/display/PSG/PSGA+Pipeline+Architecture may guide you - @Aiky30 can you also cross check with the BE input serializer and see if there any discrepancies in field naming? Thanks

**Review by cramshaw — CHANGES_REQUESTED** [review]

**Comment by cramshaw** [comment]
> Okay, I understand the reasons why this config should be in the pipeline.
> 
> @cramshaw I have a question though Should not we have a pipeline-scheme.yaml per pathogen? This would avoid handling with conflicts every time pipelines for different pathogens are edited. So I wonder whether a structure like:
> 
> ```
> config 
>     - sars_cov_2_pipeline_schema.yaml
>     - tb_pipeline_schema.yaml
> ...
> ```
> 
> could simplify things for developers / reviewers.

@pdp10 absolutely - where is the best place for this to live to allow for creation of multiple? It is definitely one config per pipeline

File: psga/config/pipeline-schema.yaml (hunk 6 above)
Comment (pdp10) [reply psga/config/pipeline-schema.yaml:89]: @cramshaw are you referring to the columns of the results.csv or something else?

File: psga/config/pipeline-schema.yaml (hunk 4 above)
Comment (Aiky30) [reply psga/config/pipeline-schema.yaml:72]: @cramshaw Shall I remove them from this file then? They also won't be on the Pathogen so I guess they can just be deleted altogether for now.

File: psga/config/pipeline-schema.yaml (hunk 5 above)
Comment (Aiky30) [reply psga/config/pipeline-schema.yaml:77]: I'll remove.

File: psga/config/pipeline-schema.yaml (hunk 6 above)
Comment (Aiky30) [reply psga/config/pipeline-schema.yaml:89]: I'll add them in. All values in the file have to have a schema definition. I'll update.

Comment (Aiky30) [reply psga/config/pipeline-schema.yaml:79]: @pdp10 in all of the example output files that I have seen, the headings are lower case. Should all field names be uppercase for input and output?

An example of what I have:
```
sample_id,pct_N_bases,pct_covered_bases,longest_no_N_run,num_aligned_reads,qc_pass,lineage,conflict,ambiguity_score,scorpio_call,scorpio_support,scorpio_conflict,scorpio_notes,version,pangolin_version,scorpio_version,constellation_version,is_designated,qc_status,qc_notes,note
```

**Comment by pdp10** [comment]
There are other pipeline parameters here: https://github.com/epi2me-labs/wf-artic/blob/a60a1e1e73bde1971c680bd3d53076127dd63fc6/nextflow_schema.json#L98 
In particular `min_len`, `max_len` and `medaka_model` should be copied from that schema to ours.

**Comment by cramshaw** [comment]
Sorry @pdp10 we don't want to include anything we aren't expecting the lab to set - this metadata file is to determine fields we want them to pass in as part of their metadata file.

**Comment by pdp10** [comment]
> Sorry @pdp10 we don't want to include anything we aren't expecting the lab to set - this metadata file is to determine fields we want them to pass in as part of their metadata file.

okay

File: psga/config/pipeline-schema.yaml (hunk 4 above)
Comment (Aiky30) [reply psga/config/pipeline-schema.yaml]: I removed this validation so assumed resolve.d

File: psga/config/pipeline-schema.yaml (hunk 5 above)
Comment (Aiky30) [reply psga/config/pipeline-schema.yaml]: Removed so now resolved.

File: psga/config/psga-schema-sars-cov-2-1.0.1.yaml [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+        slug: neb-varskip-v2
+      sequencer_technology:
+        name: ont
+        slug: ont
+      pathogen:
+        name: Sars-CoV-2
+        slug: sars-cov-2
+
+    # NEB-VarSkip_v2b
+    - name: illumina-sars-cov-2-neb-varskip-v2b
```
Comment (cramshaw) [inline psga/config/psga-schema-sars-cov-2-1.0.1.yaml:283]: We don't support all these so can we comment out those we don't have yet? (we have artic and midnight)

File: psga/config/psga-schema-sars-cov-2-1.0.1.yaml [hunk 8]
```diff
@@ -0,0 +1,476 @@
...
+        slug: neb-varskip-v2b
+      sequencer_technology:
+        name: ont
+        slug: ont
+      pathogen:
+        name: Sars-CoV-2
+        slug: sars-cov-2
+
+pipeline_version:
+  docker_tag: 1.0.1
```
Comment (cramshaw) [inline psga/config/psga-schema-sars-cov-2-1.0.1.yaml:306]: This is going to need to be templated...

File: psga/config/psga-schema-sars-cov-2-1.0.1.yaml [hunk 9]
```diff
@@ -0,0 +1,476 @@
...
+input_fields:
+  - field_name: SAMPLE_ID
+    field_type: string
+    required: true
+    validation_rules:
+    - rule_type: min_max
+      min: 36
+      max: 36
+    - rule_type: regex
+      expression: \d+
```
Comment (cramshaw) [inline psga/config/psga-schema-sars-cov-2-1.0.1.yaml:317]: does that work for - ?

Comment (Aiky30) [reply psga/config/psga-schema-sars-cov-2-1.0.1.yaml:317]: It's very loose and is checking that the value should contain numbers, a failing expression would be "^([0-9]+)$". 

I can create an expression to force a uuid pattern of x number - x number - x number ...

File: psga/config/psga-schema-sars-cov-2-1.0.1.yaml (hunk 8 above)
Comment (Aiky30) [reply psga/config/psga-schema-sars-cov-2-1.0.1.yaml:306]: What do you have in mind when referring to "templated"?

File: psga/config/psga-schema-sars-cov-2-1.0.1.yaml (hunk 7 above)
Comment (Aiky30) [reply psga/config/psga-schema-sars-cov-2-1.0.1.yaml]: Done

File: psga/config/psga-schema-sars-cov-2-1.0.1.yaml (hunk 9 above)
Comment (Aiky30) [reply psga/config/psga-schema-sars-cov-2-1.0.1.yaml]: Added a uuid pattern

**Review by cramshaw — APPROVED** [review]

File: psga/config/psga-schema-sars-cov-2.yaml [hunk 10]
```diff
@@ -0,0 +1,475 @@
...
+    - name: illumina-sars-cov-2-articv1
+      kit:
+        name: ARTIC_V1
+        slug: articv1
+      sequencer_technology:
+        name: illumina
+        slug: illumina
+      pathogen:
+        name: Sars-CoV-2
+        slug: sars-cov-2
```
Comment (pdp10) [inline psga/config/psga-schema-sars-cov-2.yaml:16]: @cramshaw how is the `name` used? The correct name is `SARS-CoV-2`, but it isn't a pipeline parameter. Is this used by the backend?

Comment (cramshaw) [reply psga/config/psga-schema-sars-cov-2.yaml:16]: this is just used by the backend - when we create the pipeline it needs to be connected in the database to a pathogen

File: psga/config/pipeline-schema.yaml (hunk 4 above)
Comment (cramshaw) [reply psga/config/pipeline-schema.yaml]: I suspect we will want to put these on pathogen - we can talk about what that should look like

**Pull request merged**